src/simulators/tensor_network/MPS.cpp

```diff
@@ -0,0 +1,398 @@
+
+	if(DEBUG) temp.print();
+	temp.apply_swap();
+	if(DEBUG) temp.print();
+	MPS_Tensor left_gamma,right_gamma;
+	rvector_t lambda;
+	MPS_Tensor::Decompose(temp, left_gamma, lambda, right_gamma);
+	left_gamma.div_Gamma_by_left_Lambda(left_lambda);
+	right_gamma.div_Gamma_by_right_Lambda(right_lambda);
+	q_reg_[index_A] = left_gamma;
+	lambda_reg_[index_A] = lambda;
+	q_reg_[index_B] = right_gamma;
+}
+
+void MPS::apply_2_qubit_gate(uint_t index_A, uint_t index_B, Gates gate_type, cmatrix_t mat)
+{
+	//for MPS
+	if(index_A + 1 < index_B)
+	{
+		apply_swap(index_A,index_B-1);
+		apply_2_qubit_gate(index_B-1,index_B, gate_type, mat);
+		apply_swap(index_A,index_B-1);
+	  return;
+	}
+	else if(index_A > index_B + 1)
+	{
+		apply_swap(index_A-1,index_B);
+		apply_2_qubit_gate(index_A,index_A-1, gate_type, mat);
+		apply_swap(index_A-1,index_B);
+		return;
+	}
+
+	bool swapped = false;
+	if(index_A >  index_B)
+	{
+		myswap<uint_t>(index_A, index_B);
+		swapped = true;
+	}
+
+	MPS_Tensor A = q_reg_[index_A], B = q_reg_[index_B];
+	rvector_t left_lambda, right_lambda;
+	//There is no lambda in the edges of the MPS
+	left_lambda  = (index_A != 0) 	    ? lambda_reg_[index_A-1] : rvector_t {1.0};
+	right_lambda = (index_B != num_qubits_-1) ? lambda_reg_[index_B  ] : rvector_t {1.0};
+
+	q_reg_[index_A].mul_Gamma_by_left_Lambda(left_lambda);
+	q_reg_[index_B].mul_Gamma_by_right_Lambda(right_lambda);
+	MPS_Tensor temp = MPS_Tensor::contract(q_reg_[index_A], lambda_reg_[index_A], q_reg_[index_B]);
+
+	if(DEBUG) temp.print();
+	switch (gate_type) {
+	case cx:
+	  temp.apply_cnot(swapped);
+	  break;
+	case cz:
+	  temp.apply_cz();
+	  break;
+	case cu:
+	{
+	  cmatrix_t Zeros = AER::Utils::Matrix::I-AER::Utils::Matrix::I;
+	  cmatrix_t temp1 = AER::Utils::concatenate(AER::Utils::Matrix::I, Zeros , 1),
+			  	temp2 = AER::Utils::concatenate(Zeros, mat, 1);
+	  cmatrix_t cu = AER::Utils::concatenate(temp1, temp2 ,0) ;
+	  temp.apply_matrix(cu);
+	  break;
+	}
+	case su4:
+	  temp.apply_matrix(mat);
+	  break;
+
+    default:
+	  throw std::invalid_argument("illegal gate for apply_2_qubit_gate"); 
+	}
+	if(DEBUG) temp.print();
+	MPS_Tensor left_gamma,right_gamma;
+	rvector_t lambda;
+	MPS_Tensor::Decompose(temp, left_gamma, lambda, right_gamma);
+	left_gamma.div_Gamma_by_left_Lambda(left_lambda);
+	right_gamma.div_Gamma_by_right_Lambda(right_lambda);
+	q_reg_[index_A] = left_gamma;
+	lambda_reg_[index_A] = lambda;
+	q_reg_[index_B] = right_gamma;
+}
+
+void MPS::change_position(uint_t src, uint_t dst)
+{
+	if(src == dst)
+		return;
+	else if(src < dst)
+		for(uint_t i = src; i < dst; i++)
+			apply_swap(i,i+1);
+	else
+		for(uint_t i = src; i > dst; i--)
+			apply_swap(i,i-1);
+}
+
+cmatrix_t MPS::Density_matrix(const reg_t &qubits) const
+{
+  // ***** Assuming ascending sorted qubits register *****
+  vector<uint_t> internalIndexes;
+  for (uint_t index : qubits)
+    internalIndexes.push_back((uint_t)index);
```

Comment:
Do we really need this casting?
If so, we don't use C-style castings, so:
```c++
internalIndexes.push_back(static_cast<uint_t>(index));
```

Reply:
Casting was not necessary. Removed it.